sandbox/cgroup/kill.go

```diff
@@ -26,89 +26,184 @@ import (
-	pids, err := pidsInFile(filepath.Join(dir, "cgroup.procs"))
-	if err != nil {
-		return err
-	}
+//
+// The freeze() callback is called exactly before killing pids is started while the thaw()
+// callback is called exactly after killing pids ends and before returning errors to give
+// a chance to recover a cgroup from a frozen state. Not propagating errors from the
+// callbacks is intentional to make it clear that they are best effort.
+var killProcessesInCgroup = func(ctx context.Context, dir string, freeze func(ctx context.Context), thaw func()) error {
+	// Keep sending SIGKILL signals until no more pids are left in cgroup
+	// to cover the case where a process forks before we kill it.
+	for {
+		pids, err := pidsInFile(filepath.Join(dir, "cgroup.procs"))
+		if err != nil {
+			return err
+		}
+		if len(pids) == 0 {
+			// no more pids
+			return nil
+		}
 
-	var firstErr error
-	for _, pid := range pids {
-		pidNotFoundErr := syscall.ESRCH
-		if err := syscallKill(pid, signal); err != nil && !errors.As(err, &pidNotFoundErr) && firstErr == nil {
-			firstErr = err
+		if freeze != nil {
+			freeze(ctx)
+		}
+		var firstErr error
+		for _, pid := range pids {
+			// This prevents a rouge fork bomb from keeping this loop running forever
+			select {
+			case <-ctx.Done():
+				return fmt.Errorf("cannot kill processes in cgroup %q: %w", dir, ctx.Err())
+			default:
+			}
+
+			pidNotFoundErr := syscall.ESRCH
+			// TODO: Use pidfs when possible to avoid killing reused pids.
+			if err := syscallKill(pid, syscall.SIGKILL); err != nil && !errors.Is(err, pidNotFoundErr) && firstErr == nil {
+				firstErr = err
+			}
+		}
+		if thaw != nil {
+			// thaw() must be called before returning to avoid keeping cgroup stuck after freeze()
+			thaw()
+		}
+		if firstErr != nil {
+			return firstErr
 		}
 	}
+}
 
-	return firstErr
+func isCgroupNotExistErr(err error) bool {
+	// fs.ErrNotExist and ENODEV are ignored in case the cgroup went away while we were
+	// processing the cgroup. ENODEV is returned by the kernel if the cgroup went
+	// away while a kernfs operation is ongoing.
+	return errors.Is(err, fs.ErrNotExist) || errors.Is(err, syscall.ENODEV)
 }
 
 func killSnapProcessesImplV1(ctx context.Context, snapName string) error {
-	if err := freezeSnapProcessesImplV1(ctx, snapName); err != nil {
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, maxKillTimeout)
+	defer cancel()
+
+	freeze := func(ctx context.Context) {
+		// This is best effort freezing, ignore all errors and continue with
+		// processes killing.
+		// This accounts for two scenarios:
+		//   - Classic snaps without a freezer cgroup
+		//   - A bug in some kernel versions where sometimes a cgroup get stuck
+		//     in FREEZING state. Given that maxKillTimeout is bigger than timeout passed to freezer
+		//     This gives a chance to thaw the cgroup and trying again.
+		ctxWithTimeout, cancel := context.WithTimeout(ctx, 1*time.Second)
```

Comment:
I would move the timeout to a variable somewhere up in the module for better visibility.